utopia-remix/app/handlers/validators.ts

```diff
@@ -1,54 +1,78 @@
-import type { AccessValidator } from '../util/api.server'
-import { ensure, getUser } from '../util/api.server'
+import type { AccessValidator, ValidationResult } from '../util/api.server'
+import { getUser, validationError, validationOk } from '../util/api.server'
 import { UserProjectPermission } from '../types'
+import { AccessLevel } from '../types'
 import type { Params } from '@remix-run/react'
 import { Status } from '../util/statusCodes'
 import { hasUserProjectPermission } from '../services/permissionsService.server'
-import { getProjectOwnerById } from '../models/project.server'
+import { getProjectOwnership } from '../models/project.server'
+import { ApiError } from '../util/errors'
 
 export function validateProjectAccess(
   permission: UserProjectPermission,
   {
-    errorMessage,
-    status,
     getProjectId,
     includeDeleted = false,
     canRequestAccess = false,
   }: {
-    errorMessage?: string
-    status?: number
     getProjectId: (params: Params<string>) => string | null | undefined
     includeDeleted?: boolean
     canRequestAccess?: boolean
   },
 ): AccessValidator {
-  return async function (req: Request, params: Params<string>) {
+  return async function (req: Request, params: Params<string>): Promise<ValidationResult> {
     const projectId = getProjectId(params)
-    ensure(projectId != null, 'project id is null', Status.BAD_REQUEST)
+    if (projectId == null) {
+      return validationError(new ApiError('Invalid project id', Status.BAD_REQUEST))
+    }
 
-    const ownerId = await getProjectOwnerById({ id: projectId }, { includeDeleted: includeDeleted })
-    ensure(ownerId != null, `Project ${projectId} not found or has no owner`, Status.NOT_FOUND)
+    const ownership = await getProjectOwnership(
+      { id: projectId },
+      { includeDeleted: includeDeleted },
+    )
+    if (ownership == null) {
+      return validationError(new ApiError('Project not found', Status.NOT_FOUND))
+    }
+
+    // the project is public, go on
+    if (ownership.accessLevel === AccessLevel.PUBLIC) {
```

Comment:
That means that every logged in user has *all permissions* to public projects (including editing them)